samza-core/src/main/java/org/apache/samza/container/ContainerHeartbeatMonitor.java

```diff
@@ -92,8 +93,13 @@ public void start() {
       if (!response.isAlive()) {
         if (isApplicationMasterHighAvailabilityEnabled) {
           LOG.warn("Failed to establish connection with {}. Checking for new AM", coordinatorUrl);
-          if (checkAndEstablishConnectionWithNewAM()) {
-            return;
+          try {
+            if (checkAndEstablishConnectionWithNewAM()) {
+              return;
+            }
+          } catch (Exception e) {
+            LOG.warn("Exception trying to connect with new AM", e);
+            throw new SamzaException(e);
```

Comment:
It will kill the thread and recreate one. I'd suggest instead to proceed with exit code as below except log the error of the exception too.

Recreating the thread doesn't guarantee that the subsequent thread can recover from the previous error and potentially run into this in a loop of some sorts without any concrete action.